### PR TITLE
Adding Profiler class

### PR DESCRIPTION
**Pull request by nupplaphil**
As a prerequisite for #6641 (see https://github.com/friendica/friendica/pull/6641#discussion_r255788812 ), we had to move the profiling data out of `App` to avoid the loopback at `DBA`.

This is done with this PR now. It was a lot of fun too because I reduced two complete log-output codings to one :-).

So what I did:
- New `Profiler` class which holds all `performance` and `callstack` information for profiling
- `Profiler` implements `ContainerInterface` for `get()` and `has()` methods of performance entries
- Every `$a->performance[]` accesor is refactored to `$a->getProfiler()->get()`
- `Profiler->saveLog()` is now writing the whole profiling data to the log
- New `ProfilerFactory` for creating the `Profiler` outside of `App`
- `DBA` is now `App` free :-)
- Created some `Profiler` tests

It is now possible to add a separate Logger (must implement `LoggerInterface`) to the `Profiler` class to avoid dumping profiling data to the normal log (or at least mark them with an own channel).

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -50,7 +50,8 @@
 		"npm-asset/fullcalendar": "^3.0.1",
 		"npm-asset/cropperjs": "1.2.2",
 		"npm-asset/imagesloaded": "4.1.4",
-		"monolog/monolog": "^1.24"
+		"monolog/monolog": "^1.24",
+		"psr/container": "^1.0"
```
Comment (MrPetovan) [inline composer.json]: Please move the PHP dependencies before the `"fxp/composer-asset-plugin"` line and sort them alphabetically.

File: src/Core/Logger.php [hunk 2]
```diff
@@ -334,7 +334,7 @@ public static function log($msg, $level = LogLevel::INFO)
 
         $stamp1 = microtime(true);
 		self::$logger->log($level, $msg);
-        self::getApp()->saveTimestamp($stamp1, "file");
+		self::getApp()->getProfiler()->saveTimestamp($stamp1, "file", System::callstack());
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Wrong indentation.

File: src/Core/Logger.php [hunk 3]
```diff
@@ -355,6 +355,6 @@ public static function devLog($msg, $level = LogLevel::DEBUG)
 
         $stamp1 = microtime(true);
         self::$devLogger->log($level, $msg);
-        self::getApp()->saveTimestamp($stamp1, "file");
+		self::getApp()->getProfiler()->saveTimestamp($stamp1, "file", System::callstack());
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Wrong indentation.

File: src/Core/Renderer.php [hunk 4]
```diff
@@ -74,7 +74,7 @@ public static function replaceMacros($s, $vars)
             exit();
         }
 
-        $a->saveTimestamp($stamp1, "rendering");
+		$a->getProfiler()->saveTimestamp($stamp1, "rendering", System::callstack());
```
Comment (MrPetovan) [inline src/Core/Renderer.php]: Wrong indentation.

File: src/Util/Profiler.php [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		$this->callstack[$value][$callstack] += (float) $duration;
+	}
+
+	/**
+	 * Resets the performance and callstack profiling
+	 *
+	 * @param bool $performance If true, reset the performance (Default true)
+	 * @param bool $callstack   If true, reset the callstack (Default true)
+	 */
+	public function reset($performance = true, $callstack = true)
```
Comment (MrPetovan) [inline src/Util/Profiler.php]: Please refactor this function into two separate `resetPerformance` and `resetCallstack` methods who will be called separately by the calling code if any parameter was `false` in the current state.

You can keep this method that will take no parameters and just call both `resetPerformance` and `resetCallstack`.

File: src/Util/Profiler.php [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+	/**
+	 * Save the current profiling data to a log entry
+	 *
+	 * @param string $message Additional message for the log
+	 */
+	public function saveLog($message = '')
+	{
+		// Write down the performance values into the log
+		if ($this->enabled) {
```
Comment (MrPetovan) [inline src/Util/Profiler.php]: Please return early instead.
```suggestion
		if (!$this->enabled) {
			return;
		}
```

File: src/Core/Logger.php (hunk 2 above)
Comment (nupplaphil) [reply src/Core/Logger.php]: The wrong indent was before, wasn't it?

File: src/Core/Logger.php (hunk 3 above)
Comment (nupplaphil) [reply src/Core/Logger.php]: The wrong indent was before, wasn't it?

File: src/Core/Renderer.php (hunk 4 above)
Comment (nupplaphil) [reply src/Core/Renderer.php]: The wrong indent was before, wasn't it?

File: src/Util/Profiler.php (hunk 5 above)
Comment (nupplaphil) [reply src/Util/Profiler.php]: done.

**Comment by nupplaphil** [comment]
Output (tested on https://friendica.philipp.info):
```console
2019-02-17 00:49:01 worker [INFO]: ID 2652085: GProbe {"module":"api","action":"call","database_read":0.022,"database_write":0.216,"cache_read":0,"cache_write":0,"network_io":95.54,"file_io":0,"other_io":6.74,"t
otal":102.52} {"file":"Profiler.php","line":163,"function":"saveLog","uid":"7de696c","process_id":17442}
2019-02-17 00:49:01 worker [INFO]: ID 2652085: GProbe:  Database Read: q, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.001 DBA::count, PortableContact::checkServer, PortableContact::las
tUpdated, GProbe::execute: 0.017 DBA::exists, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 0.001 DBA::update, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::e
xecute: 0.187 DBA::update, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.028  Database Write: DBA::update, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 0.
187 DBA::update, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.028  Cache Read:  Cache Write:  Network: PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute, Worke
r::execFunction: 70.258 Network::curl, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 12.714 PortableContact::fetchNodeinfo, PortableContact::checkServer, PortableContact::lastUpdat
ed, GProbe::execute: 8.498 PortableContact::lastUpdated, GProbe::execute, Worker::execFunction, Worker::execute: 4.073  [] {"file":"Profiler.php","line":217,"function":"saveLog","uid":"7de696c","process_id":1744
2}
2019-02-17 00:49:02 worker [INFO]: ID 2652086: GProbe {"module":"api","action":"call","database_read":0.025,"database_write":0.167,"cache_read":0.001,"cache_write":0,"network_io":88.92,"file_io":0,"other_io":8.9
1,"total":98.03} {"file":"Profiler.php","line":163,"function":"saveLog","uid":"975a8a8","process_id":17836}
2019-02-17 00:49:02 worker [INFO]: ID 2652086: GProbe:  Database Read: q, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.001 DBA::count, PortableContact::checkServer, PortableContact::las
tUpdated, GProbe::execute: 0.017 DBA::update, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 0.058 DBA::exists, Worker::entriesExists, Worker::tooMuchWorkers, Worker::add: 0.002 q, 
GContact::update, PortableContact::lastUpdated, GProbe::execute: 0.001 DBA::update, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.108  Database Write: DBA::update, PortableContact::check
Server, PortableContact::lastUpdated, GProbe::execute: 0.058 DBA::unlock, GContact::getId, GContact::update, PortableContact::lastUpdated: 0.001 DBA::update, PortableContact::lastUpdated, GProbe::execute, Worker
::execFunction: 0.108  Cache Read:  Cache Write:  Network: PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 64.706 Network::curl, PortableContact::checkServer, P
ortableContact::lastUpdated, GProbe::execute: 12.975 PortableContact::fetchNodeinfo, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 7.588 PortableContact::lastUpdated, GProbe::execu
te, Worker::execFunction, Worker::execute: 3.656  [] {"file":"Profiler.php","line":217,"function":"saveLog","uid":"975a8a8","process_id":17836}
2019-02-17 00:49:03 worker [INFO]: ID 2652102: GProbe {"module":"api","action":"call","database_read":0.024,"database_write":0.083,"cache_read":0.001,"cache_write":0,"network_io":1.3,"file_io":0,"other_io":0,"to
tal":1.42} {"file":"Profiler.php","line":163,"function":"saveLog","uid":"7de696c","process_id":17442}
2019-02-17 00:49:03 worker [INFO]: ID 2652102: GProbe:  Database Read: q, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.001 DBA::count, PortableContact::checkServer, PortableContact::las
tUpdated, GProbe::execute: 0.018 DBA::exists, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 0.001 DBA::update, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::e
xecute: 0.024 Worker::tooMuchWorkers, Worker::add, GContact::getId, GContact::update: 0.001 DBA::update, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.058  Database Write: DBA::update, P
ortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 0.024 DBA::unlock, GContact::getId, GContact::update, PortableContact::lastUpdated: 0.001 DBA::update, PortableContact::lastUpdated, GP
robe::execute, Worker::execFunction: 0.058  Cache Read:  Cache Write:  Network: PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute, Worker::execFunction: 0.954 PortableContact::fetchNode
info, PortableContact::checkServer, PortableContact::lastUpdated, GProbe::execute: 0.09 PortableContact::lastUpdated, GProbe::execute, Worker::execFunction, Worker::execute: 0.258  [] {"file":"Profiler.php","lin
e":217,"function":"saveLog","uid":"7de696c","process_id":17442}
```

File: src/Core/Logger.php (hunk 3 above)
Comment (MrPetovan) [reply src/Core/Logger.php]: Yes, and the line before as well. Can you please check the whole file?

File: src/Util/Profiler.php (hunk 5 above)
Comment (MrPetovan) [reply src/Util/Profiler.php]: My goal was to get rid the two boolean parameters. Please adapt calls to `Profiler->reset` accordingly.

File: src/Core/Logger.php (hunk 3 above)
Comment (nupplaphil) [reply src/Core/Logger.php]: btw. done for the complete file

**Pull request merged**